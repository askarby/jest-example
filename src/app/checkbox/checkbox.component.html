<div class="form">
  <input type="checkbox" [id]="id" [checked]="checked" (change)="checkedChange.emit(ref.checked)" #ref/>
  <label [for]="id">
    <svg viewBox="0,0,50,50">
      <path d="M5 30 L 20 45 L 45 5"></path>
    </svg>
  </label>
</div>
