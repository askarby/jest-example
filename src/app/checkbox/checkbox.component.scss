$size: 1.3rem;
$border-width: 0.15em;

input[type="checkbox"] {
  position: absolute;
  opacity: 0;

  &:checked + label svg path {
    stroke: green !important;
  }

  &:checked + label svg path {
    stroke-dashoffset: 0;
  }

  &:focus + label {
    transform: scale(1.03);
  }
}

input[type="checkbox"] + label {
  display: block;
  border: $border-width solid #333;
  width: $size;
  height: $size;
  border-radius: 0.3rem;
  cursor: pointer;
  transition: all .2s ease;

  &:active {
    transform: scale(1.05);
    border-radius: 0.5rem;
  }

  svg {
    pointer-events: none;
    path {
      fill: none;
      stroke: #333;
      stroke-width: 1rem;
      stroke-linecap: round;
      stroke-linejoin: round;
      stroke-dasharray: 100;
      stroke-dashoffset: 101;
      transition: all 350ms cubic-bezier(1, 0, .37, .91);
    }
  }
}
