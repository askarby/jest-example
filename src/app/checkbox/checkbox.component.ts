import { Component, EventEmitter, Input, Output } from '@angular/core';

@Component({
  selector: 'tl-checkbox',
  templateUrl: './checkbox.component.html',
  styleUrls: ['./checkbox.component.scss']
})
export class CheckboxComponent {
  @Input()
  checked: boolean;

  @Output()
  checkedChange: EventEmitter<boolean>;

  id: string;

  constructor() {
    this.checkedChange = new EventEmitter<boolean>();
    this.id =  'check-' + Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
  }
}
