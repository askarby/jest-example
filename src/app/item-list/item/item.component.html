<div class="item">
  <tl-checkbox [(checked)]="completed"></tl-checkbox>

  <!-- Edit-mode -->
  <ng-container *ngIf="isEditing; else readOnly">
    <input type="text" #task
           [value]="item.task"
           (keydown.enter)="confirm(task.value)"
           (keydown.escape)="cancel()">
    <button class="confirm" (click)="confirm(task.value)">
      <fa-icon icon="check"></fa-icon>
    </button>
    <button class="cancel" (click)="cancel()">
      <fa-icon icon="times"></fa-icon>
    </button>
  </ng-container>

  <!-- Read-only-mode -->
  <ng-template #readOnly>
    <div class="label-container" (click)="toggleMode()">
      <span class="label" [ngClass]="{striked: completed}">{{item.task}}</span>
    </div>
    <button class="delete" (click)="delete.emit(item)">
      <fa-icon icon="trash"></fa-icon>
    </button>
  </ng-template>
</div>
