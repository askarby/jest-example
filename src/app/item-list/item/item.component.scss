$completed: #737373;

@keyframes strike {
  0% {
    width: 0;
  }
  100% {
    width: 100%;
  }
}

.item {
  display: flex;
  align-content: stretch;
}

tl-checkbox {
}

.label-container {
  flex-grow: 1;

  .label {
    padding-left: 0.5rem;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;

    &.striked {
      position: relative;
      color: $completed;
    }

    &.striked:after {
      content: '';
      position: absolute;
      top: 50%;
      left: 0.5rem;
      width: 100%;
      height: 3px;
      background: $completed;
      animation-name: strike;
      animation-duration: 0.5s;
      animation-timing-function: linear;
      animation-iteration-count: 1;
      animation-fill-mode: forwards;
    }
  }
}

input {
  flex-grow: 1;
  font-size: 20px;
  border: 0;
  border-bottom: 1px solid #ccc;
  margin-left: 0.5rem;
  margin-right: 0.5rem;
  outline: 0;
}

@mixin button($color) {
  background-color: $color;
  border-radius: 0.2rem;
  border: 0;
  outline: 0;

  &:hover {
    background-color: lighten($color, 20%);
  }

  &:active {
    background-color: darken($color, 20%);
  }
}

button.confirm {
  @include button(#2ECC40);
}

button.cancel {
  margin-left: 0.2rem;
  @include button(#FF4136);
}

button.delete {
  @include button(#aeaeae);
  display: none;
}

:hover > button.delete {
  display: inline-block;
}



