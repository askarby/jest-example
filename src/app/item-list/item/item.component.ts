import { Component, EventEmitter, forwardRef, Input, OnInit, Output } from '@angular/core';
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
import { TodoListItem } from '../../todo-list-item.model';

@Component({
  selector: 'li[tl-item]',
  templateUrl: './item.component.html',
  styleUrls: ['./item.component.scss'],
})
export class ItemComponent {

  @Input()
  item: Readonly<TodoListItem>;

  @Output()
  change: EventEmitter<TodoListItem>;

  @Output()
  delete: EventEmitter<TodoListItem>;

  isEditing: boolean;

  constructor() {
    this.change = new EventEmitter<TodoListItem>();
    this.delete = new EventEmitter<TodoListItem>();

    this.isEditing = false;
  }

  confirm(task: string) {
    const modified: TodoListItem = {
      ...this.item,
      task,
    };
    this.change.emit(modified);
    this.isEditing = false;
  }

  cancel() {
    this.isEditing = false;
  }

  set completed(completed: boolean) {
    const modified: TodoListItem = {
      ...this.item,
      isCompleted: completed,
    };
    this.change.emit(modified);
  }

  get completed() {
    return this.item.isCompleted;
  }

  toggleMode() {
    this.isEditing = !this.isEditing;
  }
}
