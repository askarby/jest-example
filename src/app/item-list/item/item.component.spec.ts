import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { ItemComponent } from './item.component';
import { mockComponent } from '../../testing/mock-component.utility';
import { createTodoListItem } from '../../testing/model-factory.utility';
import { TemplateLookup } from '../../testing/template-lookup.utility';
import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';
import { CheckboxComponent } from '../../checkbox/checkbox.component';

describe('ItemComponent', () => {
  let fixture: ComponentFixture<ItemComponent>;
  let component: ItemComponent;

  let lookup: TemplateLookup;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        ItemComponent,
        mockComponent({
          selector: 'tl-checkbox',
          inputs: ['checked'],
          outputs: ['checkedChange']
        }),
        mockComponent({
          selector: 'fa-icon',
        }),
      ]
    })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ItemComponent);
    lookup = new TemplateLookup(fixture);

    component = fixture.componentInstance;
    component.item = createTodoListItem('test');

    fixture.detectChanges();
  });

  it('should be created', () => {
    expect(component).toBeTruthy();
  });

  describe('confirm', () => {
    it('should emit a change event with the item (having the modified task)', () => {
      spyOn(component.change, 'emit');
      const newTask = 'new text';
      component.confirm(newTask);
      expect(component.change.emit).toHaveBeenCalledWith({
        ...component.item,
        task: newTask,
      });
    });
  });

  describe('cancel', () => {
    it('should set edit-mode to false', () => {
      component.isEditing = true;
      component.cancel();
      expect(component.isEditing).toBe(false);
    });
  });

  describe('set completed', () => {
    it('should emit a change event with the item (having the completed-state modified)', () => {
      spyOn(component.change, 'emit');
      component.completed = true;
      expect(component.change.emit).toHaveBeenCalledWith({
        ...component.item,
        isCompleted: true,
      });
    });
  });

  describe('get completed', () => {
    it('should retrieve the completed state from the item', () => {
      expect(component.item.isCompleted).toBe(false);
    });
  });

  describe('toggleMode', () => {
    it('should be able to toggle "isEditing"-flag', () => {
      expect(component.isEditing).toBe(false);
      component.toggleMode();
      expect(component.isEditing).toBe(true);
      component.toggleMode();
      expect(component.isEditing).toBe(false);
    });
  });

  describe('Component bindings', () => {
    it('should bind the checked state to the checkbox', () => {
      const checkbox = lookup.getComponentByCss<CheckboxComponent>('tl-checkbox');
      component.item = {
        ...createTodoListItem('test'),
        isCompleted: true,
      };
      fixture.detectChanges();
      expect(checkbox.checked).toBe(true);
    });

    describe('edit mode', () => {
      let confirmButton: DebugElement;
      let cancelButton: DebugElement;

      beforeEach(() => {
        component.toggleMode();
        fixture.detectChanges();

        confirmButton = fixture.debugElement.query(By.css('button.confirm'));
        cancelButton = fixture.debugElement.query(By.css('button.cancel'));
      });

      it('should display item in edit mode', () => {
        expect(confirmButton).not.toBeNull();
        expect(cancelButton).not.toBeNull();
      });

      it('should bind value of input to task-property of item', () => {
        const input = lookup.getSingleElementByCss('input') as HTMLInputElement;
        expect(input.value).toEqual(component.item.task);
      });

      it('should invoke the confirm-method, when the confirm button is clicked', () => {
        spyOn(component, 'confirm');
        confirmButton.nativeElement.click();
        expect(component.confirm).toHaveBeenCalled();
      });

      it('should invoke the cancel-method, when the cancel button is clicked', () => {
        spyOn(component, 'cancel');
        cancelButton.nativeElement.click();
        expect(component.cancel).toHaveBeenCalled();
      });
    });

    describe('read-only mode', () => {
      let labelContainer: DebugElement;

      beforeEach(() => {
        labelContainer = fixture.debugElement.query(By.css('.label-container'));
      });

      it('should display item in read-only mode', () => {
        expect(labelContainer).not.toBeNull();
      });

      it('should display the task-property of the item, as the label', () => {
        expect(labelContainer.nativeElement.innerHTML).toContain(component.item.task);
      });

      it('should trigger edit-mode when the label is clicked', () => {
        labelContainer.nativeElement.click();
        fixture.detectChanges();
        const confirmButton = fixture.debugElement.query(By.css('button.confirm'));
        expect(confirmButton).not.toBeNull();
      });

      it('should emit delete-event when delete button is clicked', () => {
        spyOn(component.delete, 'emit');
        const deleteButton = fixture.debugElement.query(By.css('button.delete'));
        deleteButton.nativeElement.click();
        expect(component.delete.emit).toHaveBeenCalledWith(component.item);
      });
    });
  });

  describe('snapshot of markup', () => {
    it('should match the markup for a item', () => {
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });
  });
});
