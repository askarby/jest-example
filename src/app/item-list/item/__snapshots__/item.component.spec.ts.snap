// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemComponent snapshot of markup should match the markup for a item 1`] = `
<li
  change={[Function EventEmitter]}
  delete={[Function EventEmitter]}
  isEditing="false"
  item={[Function Object]}
>
  <div
    class="item"
  >
    <tl-checkbox
      ng-reflect-checked="false"
    />
    
    <div
      class="label-container"
    >
      <span
        class="label"
        ng-reflect-klass="label"
        ng-reflect-ng-class="[object Object]"
      >
        test
      </span>
    </div>
    <button
      class="delete"
    >
      <fa-icon
        icon="trash"
      />
    </button>
    
  </div>
</li>
`;
